controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,466 @@
+		return true
+	}
+
+	return false
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareSettingsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3v1alpha1.HostFirmwareSettings{}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+			}).
+		Watches(&source.Kind{Type: &metal3v1alpha1.BareMetalHost{}}, &handler.EnqueueRequestForObject{}, builder.Predicates{}).
+		Complete(r)
+}
+
+// Validate the HostFirmwareSetting Spec against the schema
+func (r *HostFirmwareSettingsReconciler) validateHostFirmwareSettings(info *rInfo, schema *metal3v1alpha1.FirmwareSchema) []error {
+
+	var errors []error
+
+	for name, val := range info.hfs.Spec.Settings {
+		// Prohibit any Spec settings with "Password"
+		if strings.Contains(name, "Password") {
+			errors = append(errors, fmt.Errorf("Cannot set Password field"))
+			break
+		}
+
+		// The setting must be in the Status
+		if _, ok := info.hfs.Status.Settings[name]; !ok {
+			errors = append(errors, fmt.Errorf("Setting %s is not in the Status field", name))
+			break
+		}
+
+		// check validity of updated value
+		if schema != nil {
+			if err := schema.ValidateSetting(name, val, schema.Spec.Schema); err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	if len(errors) > 0 {
+		return errors
+	}
+
+	return nil
+}
+
+func (r *HostFirmwareSettingsReconciler) publishEvent(request ctrl.Request, event corev1.Event) {
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", request.NamespacedName)
+	reqLogger.Info("publishing event", "reason", event.Reason, "message", event.Message)
+	err := r.Create(context.TODO(), &event)
+	if err != nil {
+		reqLogger.Info("failed to record event, ignoring",
+			"reason", event.Reason, "message", event.Message, "error", err)
+	}
+	return
+}
+
+func setCondition(generation int64, status *metal3v1alpha1.HostFirmwareSettingsStatus,
+	cond metal3v1alpha1.SettingsConditionType, newStatus metav1.ConditionStatus,
+	reason conditionReason, message string) {
+	newCondition := metav1.Condition{
+		Type:               string(cond),
+		Status:             newStatus,
+		ObservedGeneration: generation,
+		Reason:             string(reason),
+		Message:            message,
+	}
+	meta.SetStatusCondition(&status.Conditions, newCondition)
+}
+
+// Generate a name based on the schema keys which should be the same for similar hardware
+func GetSchemaName(schema map[string]metal3v1alpha1.SettingSchema) string {
+
+	// Schemas from the same vendor and model should be identical for both keys and values.
+	// Hash the keys of the map to get a identifier unique to this schema.
+	keys := make([]string, 0, len(schema))
+	for k := range schema {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	h := sha256.New()
+	h.Write([]byte(fmt.Sprintf("%v", keys)))
+	hash := fmt.Sprintf("%x", h.Sum(nil))[:8]
+
+	return "schema-" + hash
```

Comment:
IIRC we decided to put the vendor and model in the name as well as the shortened hash. Did that turn out not to be feasible?

Reply:
I thought we wouldn't have the schema data available until after inspection either?
There is something to be said for not mixing data from two different sources though, since they could change independently.